services/src/machine_learning/xgboost_training.rs

```diff
@@ -0,0 +1,115 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+use typetag::serde;
+
+use serde_json::Value;
+use snafu::{ensure, ResultExt};
+use xgboost_rs::{Booster, DMatrix};
+
+use crate::error::Result;
+use crate::machine_learning::xg_error as XgModuleError;
+use crate::machine_learning::MachineLearningFeature;
+use ordered_float::NotNan;
+
+use super::MachineLearningAggregator;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct XgboostTrainingParams {
+    pub no_data_value: f32, // FIXME: remove?
+    pub training_config: HashMap<String, String>,
+    pub feature_names: Vec<Option<String>>,
+    pub label_name: Option<String>,
+    pub aggregate_variant: MachineLearningAggregator,
+}
+
+pub fn xgb_train_model(
+    results: &mut Vec<MachineLearningFeature>,
+    training_config: HashMap<String, String>,
+) -> Result<Value> {
+    ensure!(
+        results.len() >= 2,
+        XgModuleError::error::MachineLearningMustHaveAtLeastTwoFeatures
+    );
+
+    let n_bands = results.len() - 1;
+    let lbls = results
+        .pop()
+        .expect("There should have been at least two features!");
```

Comment:
this will panic. Is it checked somewhere prior with a proper error message?

Reply:
72ee5b13

commit msg:
refactor: rename MachineLearningMustHaveAtLeastTwoFeatures -> MachineLearningMustHaveAtLeastOneFeatureAndOneLabel